fem/coefficient.hpp

```diff
@@ -123,28 +125,98 @@ class PWConstCoefficient : public Coefficient
                        const IntegrationPoint &ip);
 };
 
+namespace detail
+{
+
+/// Default implementation of deduce_type defines no aliases
+template <typename T> struct deduce_type;
+
+/// If `T` can be converted to a std::function then deduce_type defines an
+/// alias `type` to the std::function version of T
+/// This implementation and the above default are used together to enable
+/// compile time introspection. See templated FunctionCoefficient constructors
+/// for usage.
+template <typename Treturn_type, typename Tclass_type, typename... Targs>
+struct deduce_type<Treturn_type(Tclass_type::*)(Targs...) const>
+{
+   using type = std::function<Treturn_type(Targs...)>;
+};
+
+/// shorthand for typename deduce_type<>::type
+template<typename T>
+using deduce_type_t = typename deduce_type<T>::type;
+
+/// C++14 feature enabling shorthand for typename enable_if<>::type
+template<bool Condition, typename T = void>
+using enable_if_t = typename std::enable_if<Condition, T>::type;
+}
 
 /// A general C-function coefficient
 class FunctionCoefficient : public Coefficient
 {
 protected:
-   double (*Function)(const Vector &);
-   double (*TDFunction)(const Vector &, double);
+   std::function<double(const Vector &)> Function;
+   std::function<double(const Vector &, double)> TDFunction;
 
 public:
+   /// Define a time-independent coefficient from a std function
+   FunctionCoefficient(std::function<double(const Vector &)> F)
+      : Function(F)
+   { }
```

Comment:
Since the argument is passed by value, I think it is better to move-construct `Function` with `Function(std::move(F))`. This is similar to ctor 5 here: https://en.cppreference.com/w/cpp/utility/functional/function/function.